File: app/src/main/java/com/example/myapplication/ExpenseActivity.java
package com.example.myapplication;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.SearchView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class ExpenseActivity extends AppCompatActivity {

    public static final String EXTRA_TYPES = "extra_types";
    public static final String EXTRA_TITLE = "extra_title";

    private final Calendar calendar = Calendar.getInstance();
    private Calendar selectedDayCalendar = null;
    private List<String> filterTypes;
    private String sectionTitle;
    private final List<Transaction> allTransactions = new ArrayList<>();
    private final List<Row> displayRows = new ArrayList<>();
    private ExpenseDayAdapter rowAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_expense);

        View root = findViewById(R.id.root);
        ViewCompat.setOnApplyWindowInsetsListener(root, (v, insets) -> {
            Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(bars.left, bars.top, bars.right, bars.bottom);
            return insets;
        });

        filterTypes = getIntent().getStringArrayListExtra(EXTRA_TYPES);
        if (filterTypes == null) {
            filterTypes = new ArrayList<>();
            filterTypes.add("expense");
        }
        // Normalize search types to lowercase
        for (int i = 0; i < filterTypes.size(); i++) {
            filterTypes.set(i, filterTypes.get(i).toLowerCase(Locale.US));
        }

        sectionTitle = getIntent().getStringExtra(EXTRA_TITLE);
        if (sectionTitle == null) {
            sectionTitle = "Expenses";
        }

        findViewById(R.id.ivBack).setOnClickListener(v -> finish());
        View logo = findViewById(R.id.logo);
        if (logo != null) {
            logo.setOnClickListener(v -> {
                Intent intent = new Intent(this, MainActivity.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(intent);
            });
        }
        ((TextView) findViewById(R.id.tvSectionTitle)).setText(sectionTitle);

        rowAdapter = new ExpenseDayAdapter(displayRows);
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(rowAdapter);

        setupMonthNavigation();
        setupSearch();

        FloatingActionButton fabAdd = findViewById(R.id.fabAdd);
        fabAdd.setOnClickListener(v -> {
            Intent intent = new Intent(this, AddTransactionActivity.class);
            if (selectedDayCalendar != null) {
                intent.putExtra("selected_date_millis", selectedDayCalendar.getTimeInMillis());
            }
            startActivity(intent);
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadTransactions();
    }

    private void loadTransactions() {
        long[] range = getRange();
        List<Transaction> data = DatabaseHelper.getInstance(this).getTransactions(range[0], range[1]);
        
        List<Transaction> filtered = new ArrayList<>();
        for (Transaction t : data) {
            if (filterTypes.contains(t.getType().toLowerCase(Locale.US))) {
                filtered.add(t);
            }
        }
        filtered.sort((t1, t2) -> Long.compare(t2.getDate(), t1.getDate()));

        allTransactions.clear();
        allTransactions.addAll(filtered);
        applyFilter(null);
        updateDateLabel();
    }

    private void applyFilter(String query) {
        String q = query != null ? query.trim().toLowerCase(Locale.US) : "";
        List<Transaction> filtered = new ArrayList<>();
        if (q.isEmpty()) {
            filtered.addAll(allTransactions);
        } else {
            for (Transaction t : allTransactions) {
                if (t.getTitle().toLowerCase(Locale.US).contains(q) ||
                    t.getCategory().toLowerCase(Locale.US).contains(q) ||
                    t.getType().toLowerCase(Locale.US).contains(q) ||
                    t.getNote().toLowerCase(Locale.US).contains(q) ||
                    t.getAccount().toLowerCase(Locale.US).contains(q) ||
                    String.valueOf(t.getAmount()).contains(q)) {
                    filtered.add(t);
                }
            }
        }

        double total = 0;
        for (Transaction t : filtered) {
            total += t.getAmount();
        }
        ((TextView) findViewById(R.id.tvTotalExpense)).setText(String.format(Locale.US, "₹%.2f", total));

        SimpleDateFormat dayFmt = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        Map<String, List<Transaction>> grouped = new LinkedHashMap<>();
        for (Transaction t : filtered) {
            String key = dayFmt.format(new Date(t.getDate()));
            if (!grouped.containsKey(key)) {
                grouped.put(key, new ArrayList<>());
            }
            grouped.get(key).add(t);
        }

        displayRows.clear();
        for (Map.Entry<String, List<Transaction>> entry : grouped.entrySet()) {
            double dayTotal = 0;
            for (Transaction t : entry.getValue()) {
                dayTotal += t.getAmount();
            }
            displayRows.add(new Row.Header(entry.getKey(), dayTotal));
            for (Transaction t : entry.getValue()) {
                displayRows.add(new Row.Item(t));
            }
        }
        rowAdapter.notifyDataSetChanged();
    }

    private void setupSearch() {
        SearchView searchView = findViewById(R.id.searchView);
        ImageView ivSearch = findViewById(R.id.ivSearch);

        ivSearch.setOnClickListener(v -> {
            if (searchView.getVisibility() == View.GONE) {
                searchView.setVisibility(View.VISIBLE);
                findViewById(R.id.llDateNav).setVisibility(View.GONE);
                findViewById(R.id.tvSectionTitle).setVisibility(View.GONE);
                searchView.setIconified(false);
                searchView.requestFocus();
            } else {
                searchView.setVisibility(View.GONE);
                findViewById(R.id.llDateNav).setVisibility(View.VISIBLE);
                findViewById(R.id.tvSectionTitle).setVisibility(View.VISIBLE);
                searchView.setQuery("", false);
                applyFilter(null);
            }
        });

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String q) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String q) {
                applyFilter(q);
                return true;
            }
        });

        searchView.setOnCloseListener(() -> {
            searchView.setVisibility(View.GONE);
            findViewById(R.id.llDateNav).setVisibility(View.VISIBLE);
            findViewById(R.id.tvSectionTitle).setVisibility(View.VISIBLE);
            applyFilter(null);
            return false;
        });
    }

    private void setupMonthNavigation() {
        findViewById(R.id.ivBack).setOnClickListener(v -> {
            if (selectedDayCalendar != null) {
                selectedDayCalendar.add(Calendar.DAY_OF_MONTH, -1);
                calendar.setTimeInMillis(selectedDayCalendar.getTimeInMillis());
            } else {
                calendar.add(Calendar.MONTH, -1);
            }
            loadTransactions();
        });

        findViewById(R.id.ivForward).setOnClickListener(v -> {
            if (selectedDayCalendar != null) {
                selectedDayCalendar.add(Calendar.DAY_OF_MONTH, 1);
                calendar.setTimeInMillis(selectedDayCalendar.getTimeInMillis());
            } else {
                calendar.add(Calendar.MONTH, 1);
            }
            loadTransactions();
        });

        TextView tvDateRange = findViewById(R.id.tvDateRange);
        tvDateRange.setOnClickListener(v -> {
            Calendar ref = selectedDayCalendar != null ? selectedDayCalendar : calendar;
            new DatePickerDialog(this, (view, year, month, day) -> {
                selectedDayCalendar = Calendar.getInstance();
                selectedDayCalendar.set(year, month, day);
                calendar.set(year, month, day);
                loadTransactions();
            }, ref.get(Calendar.YEAR), ref.get(Calendar.MONTH), ref.get(Calendar.DAY_OF_MONTH)).show();
        });

        tvDateRange.setOnLongClickListener(v -> {
            if (selectedDayCalendar != null) {
                selectedDayCalendar = null;
                loadTransactions();
                Toast.makeText(this, "Showing full month", Toast.LENGTH_SHORT).show();
            }
            return true;
        });
    }

    private void updateDateLabel() {
        SimpleDateFormat fmt = selectedDayCalendar != null ? new SimpleDateFormat("d MMM yyyy", Locale.getDefault())
                                                           : new SimpleDateFormat("MMM yyyy", Locale.getDefault());
        Calendar ref = selectedDayCalendar != null ? selectedDayCalendar : calendar;
        ((TextView) findViewById(R.id.tvDateRange)).setText(fmt.format(ref.getTime()));
    }

    private long[] getRange() {
        Calendar day = selectedDayCalendar;
        if (day != null) {
            Calendar s = (Calendar) day.clone();
            s.set(Calendar.HOUR_OF_DAY, 0);
            s.set(Calendar.MINUTE, 0);
            s.set(Calendar.SECOND, 0);
            s.set(Calendar.MILLISECOND, 0);
            Calendar e = (Calendar) day.clone();
            e.set(Calendar.HOUR_OF_DAY, 23);
            e.set(Calendar.MINUTE, 59);
            e.set(Calendar.SECOND, 59);
            e.set(Calendar.MILLISECOND, 999);
            return new long[]{s.getTimeInMillis(), e.getTimeInMillis()};
        } else {
            Calendar s = (Calendar) calendar.clone();
            s.set(Calendar.DAY_OF_MONTH, 1);
            s.set(Calendar.HOUR_OF_DAY, 0);
            s.set(Calendar.MINUTE, 0);
            s.set(Calendar.SECOND, 0);
            s.set(Calendar.MILLISECOND, 0);
            Calendar e = (Calendar) calendar.clone();
            e.set(Calendar.DAY_OF_MONTH, calendar.getActualMaximum(Calendar.DAY_OF_MONTH));
            e.set(Calendar.HOUR_OF_DAY, 23);
            e.set(Calendar.MINUTE, 59);
            e.set(Calendar.SECOND, 59);
            e.set(Calendar.MILLISECOND, 999);
            return new long[]{s.getTimeInMillis(), e.getTimeInMillis()};
        }
    }

    public static class ExpenseDayAdapter extends RecyclerView.Adapter<ExpenseDayAdapter.VH> {

        private final List<Row> rows;
        private final SimpleDateFormat dayLabelFmt = new SimpleDateFormat("EEEE, d MMM", Locale.getDefault());
        private final SimpleDateFormat timeFmt = new SimpleDateFormat("hh:mm a", Locale.getDefault());

        public static class VH extends RecyclerView.ViewHolder {
            public VH(View view) {
                super(view);
            }
        }

        public ExpenseDayAdapter(List<Row> rows) {
            this.rows = rows;
        }

        @Override
        public int getItemViewType(int position) {
            return rows.get(position) instanceof Row.Header ? 0 : 1;
        }

        @NonNull
        @Override
        public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            int layout = viewType == 0 ? R.layout.item_expense_header : R.layout.item_expense_row;
            View view = LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);
            return new VH(view);
        }

        @Override
        public void onBindViewHolder(@NonNull VH holder, int position) {
            Row row = rows.get(position);
            if (row instanceof Row.Header) {
                Row.Header header = (Row.Header) row;
                try {
                    Date date = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).parse(header.getDateKey());
                    if (date != null) {
                        ((TextView) holder.itemView.findViewById(R.id.tvDayLabel)).setText(dayLabelFmt.format(date));
                    }
                } catch (Exception ignored) {}
                ((TextView) holder.itemView.findViewById(R.id.tvDayTotal)).setText("₹" + (int) header.getTotal());
            } else if (row instanceof Row.Item) {
                Row.Item item = (Row.Item) row;
                Transaction t = item.getTransaction();
                ((TextView) holder.itemView.findViewById(R.id.tvTitle)).setText(t.getTitle());
                ((TextView) holder.itemView.findViewById(R.id.tvCategory)).setText(t.getCategory());
                ((TextView) holder.itemView.findViewById(R.id.tvTime)).setText(timeFmt.format(new Date(t.getDate())));
                ((TextView) holder.itemView.findViewById(R.id.tvAmount)).setText("₹" + (int) t.getAmount());
            }
        }

        @Override
        public int getItemCount() {
            return rows.size();
        }
    }
}

File: app/src/main/java/com/example/myapplication/ExpenseActivity.kt
package com.example.myapplication

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SearchView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.text.SimpleDateFormat
import java.util.*

data class DayGroup(val dateKey: String, val items: List<Transaction>)

class ExpenseActivity : AppCompatActivity() {

    private val calendar = Calendar.getInstance()
    private var selectedDayCalendar: Calendar? = null
    private lateinit var filterTypes: List<String>
    private lateinit var sectionTitle: String
    private val allTransactions = mutableListOf<Transaction>()
    private val displayRows = mutableListOf<Any>() // Row.Header or Row.Item
    private lateinit var rowAdapter: ExpenseDayAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_expense)

        filterTypes = intent.getStringArrayListExtra(EXTRA_TYPES) ?: arrayListOf("expense")
        sectionTitle = intent.getStringExtra(EXTRA_TITLE) ?: "Expenses"

        findViewById<TextView>(R.id.tvSectionTitle).text = sectionTitle

        rowAdapter = ExpenseDayAdapter(displayRows)
        findViewById<RecyclerView>(R.id.recyclerView).apply {
            layoutManager = LinearLayoutManager(this@ExpenseActivity)
            adapter = rowAdapter
        }

        setupMonthNavigation()
        setupSearch()
        setupBottomNavigation()

        findViewById<FloatingActionButton>(R.id.fabAdd).setOnClickListener {
            startActivity(Intent(this, AddTransactionActivity::class.java))
        }

        loadTransactions()
    }

    private fun loadTransactions() {
        val (start, end) = getRange()
        val data = DatabaseHelper(this).getTransactions(start, end)
            .filter { it.type.lowercase() in filterTypes }
            .sortedByDescending { it.date }
        allTransactions.clear()
        allTransactions.addAll(data)
        applyFilter(null)
        updateDateLabel()
    }

    private fun applyFilter(query: String?) {
        val q = query?.trim()?.lowercase() ?: ""
        val filtered = if (q.isEmpty()) allTransactions else allTransactions.filter {
            it.title.lowercase().contains(q) ||
            it.category.lowercase().contains(q) ||
            it.type.lowercase().contains(q) ||
            it.note.lowercase().contains(q) ||
            it.account.lowercase().contains(q) ||
            it.amount.toString().contains(q)
        }
        val total = filtered.sumOf { it.amount }
        findViewById<TextView>(R.id.tvTotalExpense).text = "₹%.2f".format(total)

        val dayFmt = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        val grouped = filtered.groupBy { dayFmt.format(Date(it.date)) }
            .entries.sortedByDescending { it.key }

        displayRows.clear()
        grouped.forEach { (key, items) ->
            displayRows.add(Row.Header(key, items.sumOf { it.amount }))
            items.forEach { displayRows.add(Row.Item(it)) }
        }
        rowAdapter.notifyDataSetChanged()
    }

    private fun setupSearch() {
        val searchView = findViewById<SearchView>(R.id.searchView)
        val ivSearch = findViewById<android.widget.ImageView>(R.id.ivSearch)

        ivSearch.setOnClickListener {
            if (searchView.visibility == View.GONE) {
                searchView.visibility = View.VISIBLE
                searchView.isIconified = false
                searchView.requestFocus()
            } else {
                searchView.visibility = View.GONE
                searchView.setQuery("", false)
                applyFilter(null)
            }
        }

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(q: String?) = false
            override fun onQueryTextChange(q: String?): Boolean { applyFilter(q); return true }
        })

        searchView.setOnCloseListener {
            searchView.visibility = View.GONE
            applyFilter(null)
            false
        }
    }

    private fun setupMonthNavigation() {
        findViewById<android.widget.ImageView>(R.id.ivBack).setOnClickListener {
            selectedDayCalendar = null
            calendar.add(Calendar.MONTH, -1)
            loadTransactions()
        }
        findViewById<android.widget.ImageView>(R.id.ivForward).setOnClickListener {
            selectedDayCalendar = null
            calendar.add(Calendar.MONTH, 1)
            loadTransactions()
        }
        findViewById<TextView>(R.id.tvDateRange).setOnClickListener {
            val ref = selectedDayCalendar ?: calendar
            DatePickerDialog(this, { _, year, month, day ->
                selectedDayCalendar = Calendar.getInstance().apply { set(year, month, day) }
                calendar.set(year, month, day)
                loadTransactions()
            }, ref.get(Calendar.YEAR), ref.get(Calendar.MONTH), ref.get(Calendar.DAY_OF_MONTH)).show()
        }
    }

    private fun updateDateLabel() {
        val fmt = if (selectedDayCalendar != null) SimpleDateFormat("d MMM yyyy", Locale.getDefault())
                  else SimpleDateFormat("MMM yyyy", Locale.getDefault())
        val ref = selectedDayCalendar ?: calendar
        findViewById<TextView>(R.id.tvDateRange).text = fmt.format(ref.time)
    }

    private fun getRange(): Pair<Long, Long> {
        val day = selectedDayCalendar
        return if (day != null) {
            val s = (day.clone() as Calendar).apply { set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0) }
            val e = (day.clone() as Calendar).apply { set(Calendar.HOUR_OF_DAY, 23); set(Calendar.MINUTE, 59); set(Calendar.SECOND, 59) }
            Pair(s.timeInMillis, e.timeInMillis)
        } else {
            val s = (calendar.clone() as Calendar).apply { set(Calendar.DAY_OF_MONTH, 1); set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0) }
            val e = (calendar.clone() as Calendar).apply { set(Calendar.DAY_OF_MONTH, getActualMaximum(Calendar.DAY_OF_MONTH)); set(Calendar.HOUR_OF_DAY, 23); set(Calendar.MINUTE, 59); set(Calendar.SECOND, 59) }
            Pair(s.timeInMillis, e.timeInMillis)
        }
    }

    private fun setupBottomNavigation() {
        val navView = findViewById<BottomNavigationView>(R.id.bottomNav)
        navView.selectedItemId = when (filterTypes.firstOrNull()) {
            "income" -> R.id.nav_expense
            "togive", "toget" -> R.id.nav_debts
            else -> R.id.nav_income
        }
        navView.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.nav_all -> { startActivity(Intent(this, MainActivity::class.java)); finish(); true }
                R.id.nav_income -> {
                    if ("expense" !in filterTypes) {
                        startActivity(intentFor("expense", "Expenses", R.id.nav_income)); finish()
                    }
                    true
                }
                R.id.nav_expense -> {
                    if ("income" !in filterTypes) {
                        startActivity(intentFor("income", "Income", R.id.nav_expense)); finish()
                    }
                    true
                }
                R.id.nav_debts -> {
                    if ("togive" !in filterTypes) {
                        startActivity(intentForDebts()); finish()
                    }
                    true
                }
                else -> false
            }
        }
    }

    private fun intentFor(type: String, title: String, @Suppress("UNUSED_PARAMETER") navId: Int) =
        Intent(this, ExpenseActivity::class.java)
            .putStringArrayListExtra(EXTRA_TYPES, arrayListOf(type))
            .putExtra(EXTRA_TITLE, title)

    private fun intentForDebts() =
        Intent(this, ExpenseActivity::class.java)
            .putStringArrayListExtra(EXTRA_TYPES, arrayListOf("togive", "toget"))
            .putExtra(EXTRA_TITLE, "Debts")

    companion object {
        const val EXTRA_TYPES = "extra_types"
        const val EXTRA_TITLE = "extra_title"
    }
}

sealed class Row {
    data class Header(val dateKey: String, val total: Double) : Row()
    data class Item(val t: Transaction) : Row()
}

class ExpenseDayAdapter(private val rows: MutableList<Any>) : RecyclerView.Adapter<ExpenseDayAdapter.VH>() {

    private val dayLabelFmt = SimpleDateFormat("EEEE, d MMM", Locale.getDefault())
    private val timeFmt = SimpleDateFormat("hh:mm a", Locale.getDefault())

    class VH(view: View) : RecyclerView.ViewHolder(view)

    override fun getItemViewType(position: Int) = if (rows[position] is Row.Header) 0 else 1

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val layout = if (viewType == 0) R.layout.item_expense_header else R.layout.item_expense_row
        return VH(LayoutInflater.from(parent.context).inflate(layout, parent, false))
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        when (val row = rows[position]) {
            is Row.Header -> {
                val date = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).parse(row.dateKey)!!
                holder.itemView.findViewById<TextView>(R.id.tvDayLabel).text = dayLabelFmt.format(date)
                holder.itemView.findViewById<TextView>(R.id.tvDayTotal).text = "₹${row.total.toInt()}"
            }
            is Row.Item -> {
                holder.itemView.findViewById<TextView>(R.id.tvTitle).text = row.t.title
                holder.itemView.findViewById<TextView>(R.id.tvCategory).text = row.t.category
                holder.itemView.findViewById<TextView>(R.id.tvTime).text = timeFmt.format(Date(row.t.date))
                holder.itemView.findViewById<TextView>(R.id.tvAmount).text = "₹${row.t.amount.toInt()}"
            }
        }
    }

    override fun getItemCount() = rows.size
}
